File: Behavior/Memento/Cpp/memento.cpp
#include <iostream>
#include <vector>
#include <memory>

// Memento类存储Originator对象的状态
class Memento
{
private:
    std::string state_;

    // 通常只允许Originator访问这个状态
    friend class Originator;

public:
    Memento(const std::string &state) : state_(state) {}

    std::string GetState() const
    {
        return state_;
    }
};

// Originator类可以创建一个备忘录，并可以使用备忘录恢复其状态
class Originator
{
private:
    std::string state_;

public:
    void SetState(const std::string &state)
    {
        std::cout << "Originator: Setting state to " << state << std::endl;
        this->state_ = state;
    }

    std::unique_ptr<Memento> Save()
    {
        return std::make_unique<Memento>(state_);
    }

    void Restore(const Memento &memento)
    {
        state_ = memento.GetState();
        std::cout << "Originator: State after restoring from memento: " << state_ << std::endl;
    }
};

// Caretaker负责保存备忘录，但是不能修改备忘录的内容
class Caretaker
{
private:
    std::vector<std::unique_ptr<Memento>> mementos_;
    Originator &originator_;

public:
    Caretaker(Originator &originator) : originator_(originator) {}

    void Backup()
    {
        std::cout << "Caretaker: Saving Originator's state..." << std::endl;
        mementos_.push_back(originator_.Save());
    }

    void Undo()
    {
        if (mementos_.empty())
        {
            std::cout << "Caretaker: No mementos to undo." << std::endl;
            return;
        }
        auto memento = std::move(mementos_.back());
        mementos_.pop_back();
        std::cout << "Caretaker: Restoring state from memento..." << std::endl;
        originator_.Restore(*memento);
    }
};

int main()
{
    Originator originator;
    Caretaker caretaker(originator);

    originator.SetState("State #1");
    caretaker.Backup();

    originator.SetState("State #2");
    caretaker.Backup();

    originator.SetState("State #3");
    caretaker.Undo();

    originator.SetState("State #4");
    caretaker.Undo();

    return 0;
}

File: Behavior/Memento/Go/memento.go
package main

import "fmt"

// Memento 是一个存储原发器状态的结构体
type Memento struct {
	state string
}

// GetState 用于获取存储的状态
func (m *Memento) GetState() string {
	return m.state
}

// Originator 是原发器结构体，它的状态需要被保存和恢复
type Originator struct {
	State string
}

// NewMemento 创建一个新的备忘录，用于保存当前状态
func (o *Originator) NewMemento() *Memento {
	return &Memento{state: o.State}
}

// Restore 从备忘录恢复状态
func (o *Originator) Restore(m *Memento) {
	o.State = m.GetState()
}

// Caretaker 是管理者结构体，负责保存备忘录
type Caretaker struct {
	mementoList []*Memento
}

// AddMemento 添加备忘录到列表
func (c *Caretaker) AddMemento(m *Memento) {
	c.mementoList = append(c.mementoList, m)
}

// GetMemento 获取备忘录
func (c *Caretaker) GetMemento(index int) *Memento {
	if index < len(c.mementoList) {
		return c.mementoList[index]
	}
	return nil
}

func main() {
	originator := Originator{}
	caretaker := Caretaker{}

	originator.State = "On"
	// 保存状态
	caretaker.AddMemento(originator.NewMemento())

	// 更改状态
	originator.State = "Off"

	// 恢复到之前的状态
	originator.Restore(caretaker.GetMemento(0))
	fmt.Println("Restored State:", originator.State)
}
